post: ambient peer discovery

kaiserd opened the pull request:
This PR adds a post on ambient peer discovery.

So far, it does not cover security/anonymity properties.
I could extend this post, or refer this to a future post once we made more progress on that.

Also, the `peer exchange` is a rough sketch in its current state and the arguments for not using Rendevouz are weak (will update).

jm-clius commented on _posts/2022-04-13-ambient-peer-discovery.md:
```diff
@@ -0,0 +1,251 @@
+
+## Discovery V5
+
+[Discovery v5](https://github.com/ethereum/devp2p/blob/master/discv5/discv5.md) is Ethereum's peer discovery protocol.
```
See note above on github permalinks

jm-clius commented on _posts/2022-04-13-ambient-peer-discovery.md:
```diff
@@ -0,0 +1,251 @@
+We use Discovery V5 as an ambient peer discovery method for Waku v2 because it is decentralized, efficient, actively researched, and has web3 as its main application area.
+Discv5 also offers mitigation techniques for various attacks, which we cover later in this post.
+
+One of the main benefits of DHT-based discovery methods is offering a global view over participating nodes. This, in turn, allows sampling random sets of nodes which is important for equally distributing load.
```
```suggestion
One of the main benefits of DHT-based discovery methods is offering a global view over participating nodes. 
This, in turn, allows sampling random sets of nodes which is important for equally distributing load.
```
sembr

jm-clius commented on _posts/2022-04-13-ambient-peer-discovery.md:
```diff
@@ -0,0 +1,251 @@
+However, because the topic is popular (has a high occurrence percentage $p$), it can still be efficiently discovered.
+
+In the future, we also plan to integrate Waku v2 capability discovery, which will not only allow asking for nodes that support Waku v2,
+but asking for Waku v2 nodes supporting specific Waku v2 subprotocols like filter or store. For the store protocol we envision sub-capabilities reflecting message topics and time frames of messages.
```
```suggestion
but asking for Waku v2 nodes supporting specific Waku v2 subprotocols like filter or store.
For the store protocol we envision sub-capabilities reflecting message topics and time frames of messages.
```

jm-clius commented:
Thanks for this, @kaiserd. A valuable post! I did a quick read-through, but will do a closer reading tomorrow

kaiserd commented:
Thank you :). I added the respective permalinks.

kaiserd commented:
Part of the security/anonymity considerations would be related to the comment on sending peer lists in https://github.com/vacp2p/rfc/issues/495

oskarth commented on _posts/2022-04-13-ambient-peer-discovery.md:
```diff
@@ -0,0 +1,254 @@
+The number of hops per query is logarithmically dependent on $n$.
+Thus, random walk searching is inefficient for small percentages $p$.
+However, random walks are more resilient against attacks.
+
```
Comment above belongs here

This is the nice segue into "we need something better"

oskarth commented on _posts/2022-04-13-ambient-peer-discovery.md:
```diff
@@ -0,0 +1,254 @@
+
+## Security Considerations
+
+We will investigate security properties of these discovery mechanisms with a focus on anonymity.
```
Anonymity features seems like it could be a separate post almost

oskarth commented on _posts/2022-04-13-ambient-peer-discovery.md:
```diff
@@ -0,0 +1,254 @@
+We will investigate security properties of these discovery mechanisms with a focus on anonymity.
+DHT approaches typically allow inferring information about the querying node. Further, sending peer lists might allow inferring information about the responding node.
+
+TODO elaborate or refer to a future post.
```
In the future we'll look more at this

oskarth commented on _posts/2022-04-13-ambient-peer-discovery.md:
```diff
@@ -0,0 +1,254 @@
+For [NAT traversal](https://docs.libp2p.io/concepts/nat/), Waku v2 currently supports the port mapping protocols [UPNP](https://en.wikipedia.org/wiki/Universal_Plug_and_Play) and [NAT-PMP](https://datatracker.ietf.org/doc/html/rfc6886) / [PCP](https://datatracker.ietf.org/doc/html/rfc6887).
+
+In the future, we plan to add support for parts of [ICE](https://datatracker.ietf.org/doc/html/rfc8445), e.g. [STUN](https://datatracker.ietf.org/doc/html/rfc7350).
+We do not plan to support [TURN](https://www.rfc-editor.org/rfc/rfc5928) as it does not match our anonymity requirements.
```
Why?

oskarth commented on _posts/2022-04-13-ambient-peer-discovery.md:
```diff
@@ -0,0 +1,254 @@
+For mobile nodes, which suffer from changing IP addresses and double NAT setups, we plan using the peer exchange protocol to ask peers for more peers.
+Besides saving resources on resource restricted devices, this approach works as long as peers are in less restrictive environments.
+
+
```
Missing  conclusion section

oskarth reviewed (commented):
Nice one! Covers quite a lot. Left some comments in-line, mostly for clarity

kaiserd commented on _posts/2022-04-13-ambient-peer-discovery.md:
```diff
@@ -0,0 +1,254 @@
+categories: research
+summary:
+image: /assets/img/vac.png
+discuss:
```
Noted. I will add that after addressing the rest of the feedback.

kaiserd commented on _posts/2022-04-13-ambient-peer-discovery.md:
```diff
@@ -0,0 +1,254 @@
+TODO elaborate or refer to a future post.
+
+
+## NAT traversal
```
NAT effects all ambient peer discovery methods, so I gave it a dedicated section. Wdyt?

kaiserd commented on _posts/2022-04-13-ambient-peer-discovery.md:
```diff
@@ -0,0 +1,254 @@
+### Random Walk Discovery
+
+While discv5 is based on the Kademlia DHT, it only uses the *distributed node set* aspect of DHTs.
+So, it basically only uses the *key set* of a hash table, and not the mapping of values (items) into the hash space.
```
I wanted to emphasize that discv5 only uses one (of two) parts of a hash table structure and not the other. I will find a different wording.

kaiserd commented:
Thank you for the feedback :). I addressed it and also gave more explanation at some points.

kaiserd commented:
I added small sections for "static node lists" and "DNS-based discovery" for completeness sake.
This gives readers a more comprehensive overview and more context.

jm-clius commented on _posts/2022-04-21-ambient-peer-discovery.md:
```diff
@@ -0,0 +1,334 @@
+Various ambient peer discovery mechanisms are supported, and each is specified as a separate protocol.
+Where do these protocols fit into Waku's protocol stack?
+The P2P layer of Waku v2 builds on [libp2p gossipsub](https://github.com/libp2p/specs/blob/10712c55ab309086a52eec7d25f294df4fa96528/pubsub/gossipsub/README.md).
+Nodes participating in a gossipsub protocol manage a mesh network that is used for routing messages.
```
Would "form" perhaps be a more accurate word choice than "manage" here?

kaiserd replied:
Agreed, it read more fluidly.
I used manage, because the nodes do manage various sets of peers: `mesh`, `fanout`, `peers`.
This illustrates that mesh management and peer discovery are in different layers. Wdyt?

oskarth commented:
@kdeme @richard-ramos @D4nte any review comments?

D4nte commented on _posts/2022-04-21-ambient-peer-discovery.md:
```diff
@@ -0,0 +1,334 @@
+
+The simplest method of learning about peers in a P2P network is via static node lists.
+These can be given to nodes as start-up parameters or listed in a config-file.
+They can also be provided in a script-parseable format, e.g. the [status fleets](https://fleets.status.im/) in JSON.
```
I would not mention this link as it's not meant to be used :)
Typo if you want to keep it.
```suggestion
They can also be provided in a script-parseable format, e.g. the [Status fleets](https://fleets.status.im/) in JSON.
```

kaiserd replied:
removed the link

D4nte commented on _posts/2022-04-21-ambient-peer-discovery.md:
```diff
@@ -0,0 +1,334 @@
+These can be given to nodes as start-up parameters or listed in a config-file.
+They can also be provided in a script-parseable format, e.g. the [status fleets](https://fleets.status.im/) in JSON.
+While this method of providing bootstrap nodes is very easy to implement, it requires static peers, which introduce centralized elements.
+Also, updating static peer information introduces administrative overhead.
```
```suggestion
Also, updating static peer information introduces administrative overhead.
```
I'd be more specific and clarify that you have to change code and then release the code.
DNS-based Discovery also has administrative overhead as you have to update the DNS record.
However, with DNS Discovery you are able to separate code/release management and bootstrap node management.

D4nte commented on _posts/2022-04-21-ambient-peer-discovery.md:
```diff
@@ -0,0 +1,334 @@
+In addition to a naive DNS approach, Ethereum's DNS-based discovery introduces efficient authentication leveraging [Merkle trees](https://en.wikipedia.org/wiki/Merkle_tree).
+
+However, changing and updating the list of bootstrap nodes still requires administrative privileges because DNS records have to be added or updated.
+Also, this method of discovery still requires centralized elements.
```
I think it'd be good to mention that DNS Discovery enables pointing to FQDN handled by other parties and hence does slightly reduce the centralized aspect of it.

D4nte commented on _posts/2022-04-21-ambient-peer-discovery.md:
```diff
@@ -0,0 +1,334 @@
+A DHT can be seen as a distributed node set of which each node is responsible for a part of the hash space.
+In contrast to unstructured P2P networks, e.g. the mesh network maintained by gossipsub,
+DHTs have a global view over the node set and the hash space (assuming the participating nodes behave well).
+
+
```
Are these double line break on purpose?

kaiserd replied:
unintentional, removed :)

D4nte reviewed (approved):
Looks good, few nitpicks

kaiserd commented:
I addressed the feedback and added a small section on attacks (most of the security/privacy/anonymity related issues will be covered in a future post).

kdeme commented on _posts/2022-04-21-ambient-peer-discovery.md:
```diff
@@ -0,0 +1,377 @@
+In an eclipse attack, nodes controlled by the attacker poison the routing tables of other nodes in way that parts of the DHT become eclipsed, i.e. invisible.
+When a controlled node is asked for the next step in a path,
+it provides another controlled node as the next step,
+effectively navigating the querying node around or away from certain areas of the DHT.
```
One mitigation here is to have alpha > 1 and additionally to follow each concurrent path independently for the lookup. Not sure if you want to mention that.

kdeme commented on _posts/2022-04-21-ambient-peer-discovery.md:
```diff
@@ -0,0 +1,377 @@
+
+While discv5 based ambient peer discovery has many desirable properties, resource restricted nodes and nodes behind restrictive NAT setups cannot run discv5 satisfactory.
+With these nodes in mind, we started working on a simple *peer exchange protocol* based on ideas proposed [here](https://github.com/libp2p/specs/issues/222).
+The peer exchange protocol will allow nodes to ask peers for further peers.
```
With `further` you mean "additional", right? Not as in, more far away (Kademlia distances wise)?

kaiserd replied:
Yes :). I changed it to "additional", as you suggested, to avoid confusion.

kdeme commented on _posts/2022-04-21-ambient-peer-discovery.md:
```diff
@@ -0,0 +1,377 @@
+
+## NAT traversal
+
+For [NAT traversal](https://docs.libp2p.io/concepts/nat/), Waku v2 currently supports the port mapping protocols [UPNP](https://en.wikipedia.org/wiki/Universal_Plug_and_Play) and [NAT-PMP](https://datatracker.ietf.org/doc/html/rfc6886) / [PCP](https://datatracker.ietf.org/doc/html/rfc6887).
```
nit:
```suggestion
For [NAT traversal](https://docs.libp2p.io/concepts/nat/), Waku v2 currently supports the port mapping protocols [UPnP](https://en.wikipedia.org/wiki/Universal_Plug_and_Play) and [NAT-PMP](https://datatracker.ietf.org/doc/html/rfc6886) / [PCP](https://datatracker.ietf.org/doc/html/rfc6887).
```

kdeme reviewed (approved):
lgtm, some minor suggestions / remarks.

kaiserd commented on _posts/2022-04-21-ambient-peer-discovery.md:
```diff
@@ -0,0 +1,377 @@
+
+A first naive solution for Waku v2 discv5 discovery is
+
+* query for a random node set
```
Agreed.
I wrote `retrieve a random node set, which is achieved by querying for a set of randomly chosen node IDs`, trying to make it easier to understand for a reader that is not so familiar with DHTs.

kaiserd commented:
Thank you very much for the feedback :). I pushed a commit addressing it.

I will merge this PR to dev now. We can still proofread on dev.vac.dev.
